### ISSUE TITLE
8253/8254 PIT emulation?

### ISSUE THREAD

**Issue by zajo** [joncampbell123/dosbox-x#2945]
### Code of Conduct & Contributing Guidelines

- [X] I agree to follow the code of conduct and the contributing guidelines.

### Have you checked that no other similar question(s) already exists?

- [X] I have searched and didn't find any similar issues.

### Question

Does DOSBox-X emulate the Programmable Interval Timer (port 40h-43h)? Any limitations?

**Comment by joncampbell123** [comment]
No limitations that I am aware, and there is some PIT emulation that goes beyond DOSBox SVN even where it concerns the mode byte and the PC speaker square wave (and how it affects "Duke Nukem").

Many DOS games and applications wouldn't even work properly without PIT emulation as they often rely on timed delays using the timer interrupt or reading the counter value of PIT 0 (or in some cases, PC speaker PIT 2 even).

PC-98 mode emulates the PIT as it behaves on those systems as well, at port 71h-77h odd.

**Comment by joncampbell123** [comment]
In some cases, there are games and demoscene productions that will switch on the clock gate (but not the output) of the PC speaker so PIT 2 cycles and counts, and then use that as a timed delay.

On the PC-98 side of things, there is one game known to use PIT 2 (connected to the baud rate clock of the COM1 serial port on PC-98 systems), as a timed delay. It doesn't actually use the serial port, it just uses it's baud rate clock via PIT 2 for timed delays.

**Comment by zajo** [comment]
Thank you for the detailed explanation! I was asking because I'm having difficulties getting my old Apple ][ emulator (Appler, https://github.com/zajo/appler) to run perfectly.

On the Apple ][, to make sound you access address C030 with the desired frequency, each access flips the polarity of the speaker. The only way to control the frequency is by counting 6502 cycles (which runs at 1 MHz). To make sounds, game programmers would "simply" sprinkle individual writes to C030 throughout their code. There were also programs, most notably The Electric Duet (https://arachnoid.com/electric_duet/index.html), which managed to play two voice music. This requires extremely precisely timed access to C030.

Appler was originally written in 1990, at that point on a 10 MHz 286 it managed to run about as fast as a real Apple ][, but correctly timing C030 access was beyond hope. I few years later (I think on 386 but it might have been a low end 486) it was possible, and I got it working by reading the PIT 0 count. Appler counts 6502 cycles, and when the emulated Apple ][ program accesses C030 (and at other times, e.g. writes to video memory), the emulator spins on PIT 0 to wait the difference. This worked wonderfully on most motherboards.

And now... we have The Electric Duet running on Appler running on DOSBox-x, and advanced Apple ][ music players don't sound great. I'm guessing some kind of interference between my emulator and your emulator each trying to synchronize the emulated program with real-time. :)

Do you have any ideas? Essentially, we need PIT 0 reads to work with microsecond accuracy. Here is the synchronization routine in Appler:: https://github.com/zajo/appler/blob/develop/src/EMULATE.ASM#L349.

**Comment by zajo** [comment]
Also, I think I don't understand something, what is the relationship between the "cpu cycles" and the PIT? If a program synchronizes itself with the PIT, presumably it shouldn't be running faster when more "cpu cycles" is requested, but this doesn't seem to be the case.

**Comment by joncampbell123** [comment]
If a program is using the PIT for timing, it should use the same amount of emulated CPU time regardless of cycle count.

If a higher cycle count is making it run faster, then it's either not using the PIT, or code not related to the PIT is running faster.

What is the game in question?

**Comment by zajo** [comment]
It's not a game, but an old Apple ][ emulator I wrote back in 1990. I explained it in my previous replies, but based on your latest response it could be a bug in how I read PIT 0:

```
synch_loop:
                mov     cx,ax
                mov     al,00000100b
                cli
                out     43h,al
                jmp     $+2
                jmp     $+2
                jmp     $+2
                in      al,40h
                jmp     $+2
                jmp     $+2
                jmp     $+2
                mov     ah,al
                in      al,40h
                sti
                xchg    al,ah
                sub     cx,ax
                sub     bx,cx
                jnb     synch_loop
```

Does this look right? The goal here is to wait a specific time, with microsecond accuracy. It worked fine on a physical 386.

**Comment by joncampbell123** [comment]
I assume AX is the previous value it saved from the timer read, and this stores the last timer count it read at exit for use next time? If not, then you're saving some unknown value as the previous value you read and using that to determine how far the timer counted.

**Comment by zajo** [comment]
Yeah, you got it right. I was mostly asking whether the reading of PIT 0 would work as expected. You'll notice the multiple `jmp $+2`, these were added because on some hardware spinning like crazy continuously reading the timer didn't work perfectly (it has to spin like this, because I do need microsecond accuracy).

**Comment by joncampbell123** [comment]
Interesting, every 386, 486, and Pentium system I've tested DOSLIB on has never had any issues with reading the timer without I/O delays like that. What particular motherboard needed that?

**Comment by joncampbell123** [comment]
Back in the day though I did have a 486 system with a really slow keyboard controller. I mean slow in that reading the byte from 60h did not immediately clear the "data ready" bit, and if you used the keyboard controller to reset the CPU, the CPU had enough time to execute a good 100-400 instructions before the keyboard controller finally got around to raising the reset line. :laughing:

**Comment by zajo** [comment]
Oh, I couldn't tell you what motherboard it was, but I remember that I implemented the synchronization in the Apple ][ emulator on my own 486 and it worked wonderfully, but on a friend's computer it did not. Adding more delay fixed that, but then I found other computers on which this still didn't work.

Actually, I have no idea how this works on dosbox-x, I doubt it relies on some hardware access, you most likely rely on the host OS to read timers for you? Could this process be too slow to get microsecond accuracy?
